deluge/core/torrentmanager.py

```diff
@@ -1223,47 +1213,176 @@ def separate_keys(self, keys, torrent_ids):
+
+    def torrents_status_update(self, torrent_ids, keys, only_updated=False):
+        return self.torrents_status.torrents_status_update(torrent_ids, keys, only_updated)
+
 
-        # Get the torrent status for each torrent_id
+class TorrentsStatus(object):
+
+    def __init__(self, torrentmanager):
+        self.status_cache = {}
+        self.requests_queue = []
+        self.last_state_update_ts = 0
+        self.tm = torrentmanager
+
+    def add_request(self, session_id, d, req_torrents, keys, only_updated=False):
+        """
+        Add a status request
+
+        Args:
+            session_id (str):
+            d (Deferred): The deferred corresponding to the request
+            req_torrents (dict): The requested torrents
+            keys (list): The status keys for the torrents
+            only_updated (bool, optional): If only torrents that have changed should be returned
+
+        """
+        self.requests_queue.insert(0, StatusRequest(session_id, d, req_torrents, keys, only_updated))
+
+    def mark_dirty(self, torrent_ids):
+        """
+        Remove the torrent ids from status_dicts
+        This marks the torrent id as dirty, i.e. the torrent id
+        has available updates.
+
+        Args:
+            torrent_ids (list): The torrent ids to mark dirty
+
+        """
         for torrent_id in torrent_ids:
-            if torrent_id not in self.torrents:
-                # The torrent_id does not exist in the dict.
-                # Could be the clients cache (sessionproxy) isn't up to speed.
-                del status_dict[torrent_id]
+            for session_id in self.status_cache:
+                self.status_cache[session_id]["torrents_status"].pop(torrent_id, None)
+
+    def clear_status_requests(self, session_id):
+        """
+        Clean up any status requests registered for this session id
+
+        """
+        orig_len = len(self.requests_queue)
+        i = 0
+        while i < len(self.requests_queue):
+            if self.requests_queue[i].session_id == session_id:
+                del self.requests_queue[i]
             else:
-                status_dict[torrent_id] = self.torrents[torrent_id].get_status(torrent_keys, diff, all_keys=not keys)
-        self.status_dict = status_dict
-        d.callback((status_dict, plugin_keys))
+                i += 1
+        log.debug("Removed %d entires from status requests queue", orig_len - len(self.requests_queue))
 
-    def torrents_status_update(self, torrent_ids, keys, diff=False):
-        """Returns status dict for the supplied torrent_ids async
+    def torrents_status_updated(self, updated_ids):
+        """
+        Update with new status from libtorrent
+
+        """
+        self.last_state_update_ts = time.time()
+        self.tm.torrents_status.mark_dirty(updated_ids)
+        self.handle_torrents_status_callback()
+
+    def handle_torrents_status_callback(self, request=None):
```

Comment:
Do you think the name of this method `handle_torrents_status_callback` would read better as something like `request_torrents_status`? Looking at it's usage, it's not a specific handle/callback method.